TS5: Fix 'type-test-generator' to unblock bumping 'build-tools'

DLehenbauer opened the pull request:
The current prelease version of 'type-test-generator' contains a change to TypeOnly<T> that strips nominal typing from strings and numbers.  While seemingly more permissive, this introduced 2 type test failures in '@fluidframework/file-driver'.

I believe the error message points to the _prototype_ of this ctor fn:
```ts
export const FileSnapshotWriterClassFactory = <TBase extends ReaderConstructor>(Base: TBase) =>
  class extends Base implements ISnapshotWriterStorage {
    ...
  }
```
And the symptoms indicate that a Boolean type is causing the TypeOnly comparison to fail, although the compiler does not indicate the specific field (it just reports that the 'prototype' field is not assignable.)

Further modifying TypeOnly<T> to preserve the original typing of other primitives (boolean, bigint, and symbol) fixes this issue.  The 'null' and 'undefined' values cannot be branded (intersection returns never) and do not require special treatment.

DLehenbauer committed:
Fix 'type-test-generator'

DLehenbauer committed:
comment/spelling

CraigMacomber commented on build-tools/packages/build-tools/src/type-test-generator/compatibility.ts:
```diff
@@ -25,10 +26,13 @@ type requireAssignableTo<_A extends B, B> = true;
 	? number
 	: T extends string
 	? string
+	: T extends boolean | bigint | symbol
```
Including number string here would actually make a difference. If we have a type like `number | string` right now it fill fall through to the object case below, but we could fix that and allow it to keep that type information by including it here.

DLehenbauer replied:
This didn't work the way I expected.

'TypeOnly<string | number>' appears to be decomposed and processed as 'TypeOnly<string> | TypeOnly<number>', which hits the earlier cases that strip nominal typing.

(I added a test cases below to illustrate / verify.)

tylerbutler commented on build-tools/packages/build-tools/src/type-test-generator/compatibility.ts:
```diff
@@ -3,6 +3,7 @@
  * Licensed under the MIT License.
  */
 
+/* eslint-disable @typescript-eslint/no-namespace */
```
Please add a note why namespaces are ok in this file, since we discourage them in general.

DLehenbauer replied:
Done.  (The answer is that they are non-exported namespaces to help organize/delimit inline compile-time tests.)

CraigMacomber commented on build-tools/packages/build-tools/src/type-test-generator/compatibility.ts:
```diff
@@ -163,3 +162,37 @@ namespace Test7 {
+	// (Nominal typing is erased.)
+	type brandedNumber = number & { brand: "Number" };
+	type _check_number1 = requireAssignableTo<TypeOnly<brandedNumber>, number>;
+	type _check_number2 = requireAssignableTo<number, TypeOnly<brandedNumber>>;
```
this branding isn't nominal typing: its actually structural. Two different declarations of this same type (ex: in two versions of the same package) are assignable to eachother.

Thus ideally TypeOnly would preserve these brands. The fact that this type brand is getting lost is a limitation of TypeOnly, and not a feature we want to ensure doesn't break: ideally, we would preserve the brand.

The above case, _check_number1 ensures TypeOnly doesn't forget the value is a number: that is good.

```suggestion
   // Due to limitations of the current version of TypeOnly, brands on primitives are lost:
	type _check_number2 = requireAssignableTo<number, TypeOnly<brandedNumber>>;
```

CraigMacomber commented on build-tools/packages/build-tools/src/type-test-generator/compatibility.ts:
```diff
@@ -163,3 +162,37 @@ namespace Test7 {
+	type _check_string2 = requireAssignableTo<string, TypeOnly<brandedString>>;
+
+	// Other primitive types are preserved as-is.
+	type brandedBoolean = boolean & { brand: "Boolean" };
```
add a test in here for a union, like `string | number | boolean |  BigInt | symbol | undefined | null` and confirm it is preserved.

DLehenbauer replied:
I added a few test variations involving unions.

DLehenbauer committed:
Update build-tools/packages/build-tools/src/type-test-generator/compatibility.ts

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

DLehenbauer committed:
more union tests

DLehenbauer commented:
Thanks @tylerbutler !

@CraigMacomber - I think I've addressed your feedback (modulo 'string | number' not working as we guessed).  LMK if you have more thoughts.
